graphsync/main.go

```diff
@@ -0,0 +1,340 @@
+
+	host, peers := makeHost(ctx, runenv, initCtx)
+	defer host.Close()
+
+	var (
+		// make datastore, blockstore, dag service, graphsync
+		ds     = dss.MutexWrap(datastore.NewMapDatastore())
+		bs     = blockstore.NewBlockstore(ds)
+		dagsrv = merkledag.NewDAGService(blockservice.New(bs, offline.Exchange(bs)))
+		gsync  = gsi.New(ctx, gsnet.NewFromLibp2pHost(host), makeLoader(bs), makeStorer(bs))
+	)
+
+	switch runenv.TestGroupID {
+	case "providers":
+		if runenv.TestGroupInstanceCount > 1 {
+			panic("test case only supports one provider")
+		}
+
+		runenv.RecordMessage("we are the provider")
+		defer runenv.RecordMessage("done provider")
+
+		return runProvider(ctx, runenv, initCtx, dagsrv, size, latencies, bandwidths)
+
+	case "requestors":
+		runenv.RecordMessage("we are the requestor")
+		defer runenv.RecordMessage("done requestor")
+
+		p := *peers[0]
+		if err := host.Connect(ctx, p); err != nil {
+			return err
+		}
+		runenv.RecordMessage("done dialling provider")
+		return runRequestor(ctx, runenv, initCtx, gsync, p, bs, latencies, bandwidths)
+
+	default:
+		panic("unsupported group ID")
+	}
+}
+
+func runRequestor(ctx context.Context, runenv *runtime.RunEnv, initCtx *run.InitContext, gsync gs.GraphExchange, p peer.AddrInfo, bs blockstore.Blockstore, latencies []time.Duration, bandwidths []uint64) error {
+	// create a selector for the whole UnixFS dag
+	ssb := builder.NewSelectorSpecBuilder(basicnode.Style.Any)
+	sel := ssb.ExploreRecursive(
+		selector.RecursionLimitNone(),
+		ssb.ExploreAll(
+			ssb.ExploreRecursiveEdge()),
+	).Node()
+
+	for i, latency := range latencies {
+		for j, bandwidth := range bandwidths {
+			round := i*len(latencies) + j
+
+			var (
+				topicCid  = sync.NewTopic(fmt.Sprintf("cid-%d", round), new(cid.Cid))
+				stateNext = sync.State(fmt.Sprintf("next-%d", round))
+				stateNet  = sync.State(fmt.Sprintf("network-configured-%d", round))
+			)
+
+			runenv.RecordMessage("waiting to start round %d", round)
+			initCtx.SyncClient.MustSignalAndWait(ctx, stateNext, runenv.TestInstanceCount)
+
+			sctx, scancel := context.WithCancel(ctx)
+			cidCh := make(chan *cid.Cid, 1)
+			initCtx.SyncClient.MustSubscribe(sctx, topicCid, cidCh)
+			cid := <-cidCh
+			scancel()
+
+			// make a go-ipld-prime link for the root UnixFS node
+			clink := cidlink.Link{Cid: *cid}
+
+			runenv.RecordMessage("ROUND %d: latency=%s, bandwidth=%d", round, latency, bandwidth)
+			runenv.RecordMessage("CID: %s", cid)
+
+			runenv.RecordMessage("waiting for provider's network to be configured %d", round)
+			<-initCtx.SyncClient.MustBarrier(ctx, stateNet, 1).C
+			runenv.RecordMessage("network configured for round %d", round)
+
+			// execute the traversal.
+			runenv.RecordMessage(">>>>> requesting")
+			progressCh, errCh := gsync.Request(ctx, p.ID, clink, sel)
+			for r := range progressCh {
+				runenv.RecordMessage("******* progress: %+v", r)
+			}
+
+			runenv.RecordMessage("<<<<< request complete")
+			if len(errCh) > 0 {
```

Comment:
this is possibly poor design on graphsync's part, but this will not work as a way to determine if there are errors.
graphsync collects responses and errors in an array buffer so they can be read asynchronously by caller of request at any time. (https://github.com/ipfs/go-graphsync/blob/master/requestmanager/responsecollector.go) A side affect is the len() on a channel will always be zero.

you need to iterate over the error channel to determine if there are errors. Don't worry -- iterating over the response and error channels can be done independently without blocking the other. if you iterate over the error channel after already iterating on the progress channel, it will not block at all.